### PR TITLE
feat: settings

### LINKED ISSUE #62: implement getSettings with nim-sqlcipher
https://github.com/status-im/status-go/blob/4ecca1169d2e0a07b4e0cf5a7a3f4044216c678f/multiaccounts/accounts/database.go

### PR DESCRIPTION
**Pull request by richard-ramos**
fixes #62

### CODE REVIEW

File: nim_status/lib/settings/utils.nim
Comment (emizzle) [inline nim_status/lib/settings/utils.nim:43]: A generic alternative (to reduce boilerplate) might be:
```nim
proc getOption[T](self: JsonNode, key: string): Option[T] =
  if not self.hasKey(key) or self.to(T) == T():
    result = none(T)
  else:
    # handle special cases
    when T is Address:
      result = some(parseAddress(self[key].to(T))
    when T is int64
      result = some(self[key].getBiggestInt)
    # for all other (default) cases
    else:
      result = some(self[key].to(T))
```

File: nim_status/lib/settings/types.nim
Comment (emizzle) [inline nim_status/lib/settings/types.nim:110]: It is my understanding that `{.serializedFieldName.}` is only needed if `json-serialization` is used, unless there is something I'm missing.

Comment (emizzle) [reply nim_status/lib/settings/types.nim:110]: Another way to prevent string mistakes would be use the stringified enum values here, like so:
```nim
chaosMode* {.serializedFieldName($SettingsEnum.ChaosMode).}: Option[bool]
```

**Review by emizzle — COMMENTED** [review]
Awesome work! Left a few comments on adding json-serialization haha.

**Review by emizzle — APPROVED** [review]
Very well done! 👏

File: nim_status/lib/settings.nim
Comment (emizzle) [inline nim_status/lib/settings.nim:245]: There is probably a very clever way of deserialising a `DbValue` in to a type that would prevent the need for this.

```nim
proc to*(dbVal: DbValue, val: var bool) =
  val = dbVal.boolVal

proc to*(dbVal: DbValue, val: var int) =
  val = dbVal.intVal

proc to*(dbVal: DbValue, val: var string) =
  val = dbVal.strVal

# add more overloads for other types

proc to[T]*(dbVal: DbValue, t: typedesc[T]): T
  if fieldPairs(default(T)).len == 0:
    dbVal.to(result) # should go to one of the above proc overloads
    return result
  result = T()
  let colIdx = 0
  for key, val in fieldPairs(result):
    val = dvVal[colIdx].to(val.type) # need to make sure var in fieldPairs is mutable (I don't think it is)

for r in rows(db, query):
  result = r.to(Settings)
  break
```

Comment (emizzle) [reply nim_status/lib/settings.nim:245]: Btw, would the `break` only set the first row value? Maybe that's the intention?

File: nim_status/lib/settings.nim
Comment (jrainville) [inline nim_status/lib/settings.nim:195]: Wouldn't a `SELECT *` be simpler here?

File: nim_status/lib/settings/types.nim
Comment (jrainville) [inline nim_status/lib/settings/types.nim:59]: This is the same as the enum with have in `nim-status-client` right? I guess once this is merged, we could use this `enum` instead of the one in the client repo

File: nim_status/lib/settings.nim
Comment (richard-ramos) [reply nim_status/lib/settings.nim:195]: Yes, but then I do not know the order of the columns in order to populate the object attributes. (I could not find in sqlite docs if it's possible to refer to column names in the results of the query)

File: nim_status/lib/settings/types.nim
Comment (richard-ramos) [reply nim_status/lib/settings/types.nim:59]: Yeah!

File: nim_status/lib/settings.nim
Comment (richard-ramos) [reply nim_status/lib/settings.nim:245]: Yeah!

Comment (richard-ramos) [reply nim_status/lib/settings.nim:245]: > There is probably a very clever way of deserialising a `DbValue` in to a type that would prevent the need for this.

I'll try that approach, it looks neat!

**Comment by michaelsbradleyjr** [comment]
This PR branch has been rebased against latest changes to this repo's [`feat/sqlcipher`](#61) PR branch

**Comment by michaelsbradleyjr** [comment]
The test added by this PR is failing on Windows, but only when ssl, etc. are linked as shared libraries. I'm re-running the tests to check if it fails consistently in that scenario. I don't have a theory yet as to why it fails, but I also created PR #97 to check whether some bumps in `vendor/` that come with the waku v2 PR might sort it out.

For now I've added a *don't merge yet* label.

**Comment by michaelsbradleyjr** [comment]
> but I also created PR #97 to check whether some bumps in `vendor/` that come with the waku v2 PR might sort it out

PR #97 fails in the same way:

```
[NimScript] exec: test/nim/build/settings
D:\a\nim-status\nim-status\test\nim\settings.nim(33) settings
D:\a\nim-status\nim-status\nim_status\lib\settings.nim(199) getSettings
D:\a\nim-status\nim-status\nim_status\lib\settings\utils.nim(14) toOption
D:\a\nim-status\nim-status\vendor\nimbus-build-system\vendor\Nim\lib\system\fatal.nim(49) sysFatal
Error: unhandled exception: 'intVal' is not accessible using discriminant 'kind' of type 'DbValue' [FieldError]
```

**Comment by michaelsbradleyjr** [comment]
This PR's branch has been reset with the commits on #97's branch.

**Comment by michaelsbradleyjr** [comment]
Rebased against `master` and force-pushed again.

**Pull request merged**
